source/game/field/BoxColManager.cc

```diff
@@ -0,0 +1,620 @@
+    if (m_unitCount >= MAX_UNIT_COUNT) {
+        return nullptr;
+    }
+
+    s32 unitID = m_nextUnitID;
+    BoxColUnit *pUnit = &m_unitPool[unitID];
+    pUnit->init(radius, maxSpeed, pos, flag, userData);
+    m_nextUnitID = m_unitIDs[unitID];
+    f32 range = radius + maxSpeed;
+    f32 zHigh = pos->z + range;
+    f32 zLow = pos->z - range;
+
+    if (m_unitCount == 0) {
+        m_highPoints[0].lowPoint = 0;
+        m_highPoints[0].minLowPoint = 0;
+        m_lowPoints[0].unitID = unitID;
+        m_highPoints[0].z = zHigh;
+        m_lowPoints[0].highPoint = 0;
+        m_lowPoints[0].unitID = unitID;
+        m_lowPoints[0].z = zLow;
+        m_unitPool[0].m_highPointIdx = 0;
+        m_unitPool[0].m_lowPointIdx = 0;
+        m_unitCount = 1;
+
+        return pUnit;
+    }
+
+    // Binary search
+    int highPointIdx = 0;
+    int lowPointIdx = 0;
+    int i = m_unitCount;
+
+    do {
+        int highSearch = highPointIdx + i;
+        int lowSearch = lowPointIdx + i;
+
+        if (highSearch <= m_unitCount && zHigh > m_highPoints[highSearch - 1].z) {
+            highPointIdx = highSearch;
+        }
+
+        if (lowSearch <= m_unitCount && zLow > m_lowPoints[lowSearch - 1].z) {
+            lowPointIdx = lowSearch;
+        }
+
+        i = (i + 1) / 2;
+    } while (i > 1);
+
+    pUnit->m_highPointIdx = highPointIdx;
+    pUnit->m_lowPointIdx = lowPointIdx;
+
+    // Update high points
+    for (int i = m_unitCount; i > highPointIdx; --i) {
+        BoxColHighPoint &high = m_highPoints[i];
+        m_highPoints[i] = m_highPoints[i - 1];
+        BoxColLowPoint &low = m_lowPoints[high.lowPoint];
+
+        ++low.highPoint;
+        ++m_unitPool[low.unitID].m_highPointIdx;
+
+        if (high.minLowPoint >= lowPointIdx) {
+            ++high.minLowPoint;
+        }
+    }
+
+    m_highPoints[highPointIdx].lowPoint = lowPointIdx;
+    m_highPoints[highPointIdx].z = zHigh;
+
+    // Update min low point
+    if (highPointIdx == m_unitCount || m_highPoints[highPointIdx + 1].minLowPoint > lowPointIdx) {
+        m_highPoints[highPointIdx].minLowPoint = lowPointIdx;
+
+        for (int i = highPointIdx - 1; i >= 0 && m_highPoints[i].minLowPoint > lowPointIdx; --i) {
+            m_highPoints[i].minLowPoint = lowPointIdx;
+        }
+    } else {
+        m_highPoints[highPointIdx].minLowPoint = m_highPoints[highPointIdx + 1].minLowPoint;
+    }
+
+    // Update low points
+    for (int i = m_unitCount; i > lowPointIdx; --i) {
+        BoxColLowPoint &low = m_lowPoints[i];
+        m_lowPoints[i] = m_lowPoints[i - 1];
+        BoxColHighPoint &high = m_highPoints[low.highPoint];
+        ++high.lowPoint;
+        ++m_unitPool[low.unitID].m_lowPointIdx;
+    }
+
+    m_lowPoints[lowPointIdx].highPoint = highPointIdx;
+    m_lowPoints[lowPointIdx].unitID = unitID;
+    m_lowPoints[lowPointIdx].z = zLow;
+    ++m_unitCount;
+    return pUnit;
+}
+
+/// @addr{0x80786578}
+void BoxColManager::remove(BoxColUnit **unit) {
```

Comment:
```suggestion
void BoxColManager::remove(BoxColUnit *&pUnit) {
```
This avoids constantly wrapping `(*unit)` whenever we need to operate on it.